ExecHashMap.hpp

```diff
@@ -0,0 +1,119 @@
+#ifndef EXEC_HASH_MAP_HPP
+#define EXEC_HASH_MAP_HPP
+
+#include <iostream>
+#include <iterator>
+#include <numeric>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include <nlohmann/json.hpp>
+
+typedef std::vector<std::string> IOVector;
+typedef std::pair<IOVector, IOVector> IOPair;
+
+/**
+ * @brief A hash map that maps inputs to a vector of outputs
+ * @details Key is inputs (a vector), value is a vector of outputs
+ * (a vector of vectors) of multiple executions
+ */
+class ExecHashMap {
+private:
+  struct VectorHasher {
+    // https://stackoverflow.com/questions/20511347/a-good-hash-function-for-a-vector
+    // XOR of each string's hash value
+    std::size_t operator()(const std::vector<std::string> &V) const {
+      std::hash<std::string> hasher;
+      std::size_t seed = 0;
+      for (const auto &str : V) {
+        seed ^= hasher(str) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+      }
+      return seed;
+    }
+  };
+
+  std::unordered_map<IOVector, std::vector<IOVector>, VectorHasher> map;
+  int value_capacity;
+
+public:
+  ExecHashMap() : ExecHashMap(0) {}
+
+  /**
+   * @brief Construct a new Exec Hash Map object
+   * @param cap the capacity of the value vector (maxmium length)
+   */
+  ExecHashMap(int cap) : value_capacity(cap) {}
+
+  /**
+   * @brief Insert a pair of inputs and outputs to the hash map
+   * @param io: a pair of inputs (vector) and outputs (vector)
```

Comment:
What's a pair of inputs? Do you mean multiple inputs and their corresponding outputs?